[inductor] Implement Fx graph caching to improve warm compilation time.

ani300 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #103453

Summary: Implement an on-disk cache to save and reuse compiled FX Graphs. This implementation does not handle tensors with symbolic shapes. This needs to be done in a follow-up PR.

Test Plan:
* New unit tests exercising saving and load from the cache.
* New unit tests to exercise the cache key calculations.
* Ran several benchmarks to see cache hit and resulting compilation times.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler

Differential Revision: [D50255289](https://our.internmc.facebook.com/intern/diff/D50255289)

ani300 committed:
Add codegen cache

[ghstack-poisoned]

ani300 committed:
Update on "Add codegen cache"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 aakhundov

[ghstack-poisoned]

ani300 committed:
Update on "Add codegen cache"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 aakhundov

[ghstack-poisoned]

ani300 commented:
Please see https://github.com/pytorch/pytorch/pull/101027 for previous discussion on this PR

ani300 committed:
Update on "Add codegen cache"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 aakhundov

[ghstack-poisoned]

ani300 committed:
Update on "Add codegen cache"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 aakhundov

[ghstack-poisoned]

ani300 committed:
Update on "Add codegen cache"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 aakhundov

[ghstack-poisoned]

ezyang commented on torch/_inductor/codecache.py:
```diff
@@ -295,6 +329,52 @@ def write_atomic(path: str, content: Union[str, bytes]):
+                    disk_compiled_graph = copy(compiled_graph)
+                    # Important as compiled models are not pickeable
+                    disk_compiled_graph.compiled_artifact = None
+                    print(disk_compiled_graph)
```
Print

ani300 replied:
removed

ezyang commented:
If we add some tests, I'm willing to land this at let people kick the tires a bit.

One long term design thing we need to figure out is how to make sure people add extra things to the cache key when necessary; it is easy to forget.

ani300 committed:
Update on "Add codegen cache"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 aakhundov

[ghstack-poisoned]

ani300 committed:
Update on "Add codegen cache"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 aakhundov

[ghstack-poisoned]

ani300 commented:
@msaroufim @jansel all the comments on #101027 have been applied to this split PR.

@ezyang I hope to have the tests ready by the end of this week

ani300 committed:
Update on "Add codegen cache"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 aakhundov

[ghstack-poisoned]

ani300 committed:
Update on "Add codegen cache"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 aakhundov

[ghstack-poisoned]

ani300 commented:
@eellison I am having some trouble with a unit test that involves cuda graph trees and a _dynamo.reset(), where after the reset the model will refuse to compile due to cuda graphs being shut down. Do you have any suggestions on how to get it to run?

ezyang commented:
I think I need more details, unfortunately

ani300 commented:
@ezyang On the new `test_cache.py` file, the test `test_cache_cudagraphs` fails because running a new model after `torch._dynamo.reset()` with the CUDA graph trees activated triggers an assert. I don't know if this is a bug or expected behavior, but my expectation would be that I can run CUDA Graph Trees just fine after a `reset()`. The need for a reset is so that the new codegen cache activates and loads the code from disk instead of whatever is already on memory.

ezyang commented on torch/_inductor/codecache.py:
```diff
@@ -308,6 +361,59 @@ def write_atomic(path: str, content: Union[str, bytes]):
+        fx_kwargs.pop("graph_id")
+        # Add arguments to be hashed
+        fx_args_for_hashing = copy(fx_args)
+        fx_args_for_hashing.extend(list(fx_kwargs.values()))
```
nit: list here is not necessary
 
More substantively, taking kwargs this way is only OK assuming you always pass in the kwargs in the same order (which is the case at the call site, but may be non-obvious if there ever are other call sites)

ezyang commented on torch/_inductor/codecache.py:
```diff
@@ -308,6 +361,59 @@ def write_atomic(path: str, content: Union[str, bytes]):
+                    )
+
+                    disk_compiled_graph = copy(compiled_graph)
+                    # Important as compiled models are not pickeable
```
pickle'able

ezyang commented on torch/_inductor/codecache.py:
```diff
@@ -268,18 +272,67 @@ def code_hash(code, extra=""):
+def serialize_compiled_fx_arg(arg: Any):
+    if isinstance(arg, torch.fx.GraphModule):
+        # Special case for fx_graph
+        return str(arg).encode("utf-8")
```
This is a bad way to hash the graph, we should have a TODO to make it better ;)

Note that @jansel in compiled backwards had to do something similar, although I don't think any of the code will be reusable here.

ezyang commented on test/inductor/test_cache.py:
```diff
@@ -0,0 +1,150 @@
+
+
+def activate_unique_cache():
+    new_cache_key = random.randint(0, 1000000000)
```
IMO, it would be better not to have any randomness here, and just rely on clearing out the cache appropriately, since it will avoid nondeterministic tests

ezyang commented on test/inductor/test_cache.py:
```diff
@@ -0,0 +1,150 @@
+    def setUp(self):
+        super().setUp()
+        # We don't want a fixed seed for this test to avoid cache collisions
+        set_rng_seed(int(time.time()))
```
same here, really try to make the tests deterministic, and figure out some other way to deal with collisions.

ezyang commented on test/inductor/test_cache.py:
```diff
@@ -0,0 +1,150 @@
+            # The cached_result coming from a cached kernel should be the same as the original one
+            # The compilation metrics of the cached compilation should show no codegen
+            self.assertEqual(base_result, cached_result)
+            self.assertNotIn("Scheduler.codegen", cached_compilation_metrics)
```
One of the more important classes of tests here will be ensuring that hash keys don't falsely match when they shouldn't. It will be good to think of a good strategy for this.

ani300 committed:
Update on "Add codegen cache"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 aakhundov

[ghstack-poisoned]

ani300 committed:
Update on "Add codegen cache"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 aakhundov

[ghstack-poisoned]

ezyang commented:
holler when you'r eready for rereview

ani300 commented:
will do! just getting back into things right now, but I still can't get the CUDAGraphs to compile after a dynamo.reset() in a unit test...

voznesenskym commented:
> will do! just getting back into things right now, but I still can't get the CUDAGraphs to compile after a dynamo.reset() in a unit test...

Do you need a hand?

ani300 commented:
@voznesenskym from my message a few months ago:

> On the new test_cache.py file, the test test_cache_cudagraphs fails because running a new model after torch._dynamo.reset() with the CUDA graph trees activated triggers an assert. I don't know if this is a bug or expected behavior, but my expectation would be that I can run CUDA Graph Trees just fine after a reset(). The need for a reset is so that the new codegen cache activates and loads the code from disk instead of whatever is already on memory.

msaroufim commented:
Cc @eellison who might also know what's going on

ezyang commented:
Oh, now that I think about this, it might be because dynamo reset isn't resetting the cuda graph trees manager

ani300 commented:
@ezyang it looks like it does (see https://github.com/pytorch/pytorch/blob/3f2ecf7755935bce603fdf8e0537d6a57e80864c/torch/_inductor/cudagraph_trees.py#L309) but somehow it doesn't reinitialize correctly?

eellison commented:
As you someone linked to, when torch._dynamo.reset() is called, it will invalidate all cudagraphs and clear the memory: https://github.com/pytorch/pytorch/blob/main/torch/_inductor/cudagraph_trees.py#L309-L325

If you reset dynamo and therefore reset cudagraphs, and then later load the function from cache, you'll need to re-initialize and re-cudagraph it.

ani300 commented:
@eellison This code will fail, what would be missing?

```
def test_cudagraphs_error():
    model1 = Foo()
    
    cuda_model1 = model.cuda()
    opt_model1 = torch.compile(cuda_model1, mode="reduce-overhead")
    
    input = torch.rand(15, 15).cuda()
    
    result1 = opt_model1(input)
    
    torch._dynamo.reset()
    
    cuda_model2 = Foo().cuda()
    opt_model2 = torch.compile(cuda_model2, mode="reduce-overhead")
    result2 = opt_model2(input)
```

eellison commented:
Could you post error stack ?

ani300 committed:
Update on "Add codegen cache"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 aakhundov

[ghstack-poisoned]

ani300 commented:
@eellison 
```
Traceback (most recent call last):
  File "/dccstor/aviros_pytorch/pytorch/test/inductor/test_cache.py", line 142, in test_cache_cudagraphs
    cached_result = opt_model2(input2)
  File "/dccstor/aviros_pytorch/pytorch/torch/nn/modules/module.py", line 1518, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
  File "/dccstor/aviros_pytorch/pytorch/torch/nn/modules/module.py", line 1527, in _call_impl
    return forward_call(*args, **kwargs)
  File "/dccstor/aviros_pytorch/pytorch/torch/_dynamo/eval_frame.py", line 333, in _fn
    return fn(*args, **kwargs)
  File "/dccstor/aviros_pytorch/pytorch/torch/nn/modules/module.py", line 1518, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
  File "/dccstor/aviros_pytorch/pytorch/torch/nn/modules/module.py", line 1527, in _call_impl
    return forward_call(*args, **kwargs)
  File "/dccstor/aviros_pytorch/pytorch/test/inductor/test_cache.py", line 24, in forward
    def forward(self, x: torch.Tensor):
  File "/dccstor/aviros_pytorch/pytorch/torch/_dynamo/eval_frame.py", line 333, in _fn
    return fn(*args, **kwargs)
  File "/dccstor/aviros_pytorch/pytorch/torch/_dynamo/external_utils.py", line 17, in inner
    return fn(*args, **kwargs)
  File "/dccstor/aviros_pytorch/pytorch/torch/_functorch/aot_autograd.py", line 3905, in forward
    return compiled_fn(full_args)
  File "/dccstor/aviros_pytorch/pytorch/torch/_functorch/aot_autograd.py", line 1482, in g
    return f(*args)
  File "/dccstor/aviros_pytorch/pytorch/torch/_functorch/aot_autograd.py", line 2527, in runtime_wrapper
    all_outs = call_func_with_args(
  File "/dccstor/aviros_pytorch/pytorch/torch/_functorch/aot_autograd.py", line 1506, in call_func_with_args
    out = normalize_as_list(f(args))
  File "/dccstor/aviros_pytorch/pytorch/torch/_functorch/aot_autograd.py", line 1482, in g
    return f(*args)
  File "/dccstor/aviros_pytorch/pytorch/torch/autograd/function.py", line 539, in apply
    return super().apply(*args, **kwargs)  # type: ignore[misc]
  File "/dccstor/aviros_pytorch/pytorch/torch/_functorch/aot_autograd.py", line 3010, in forward
    fw_outs = call_func_with_args(
  File "/dccstor/aviros_pytorch/pytorch/torch/_functorch/aot_autograd.py", line 1506, in call_func_with_args
    out = normalize_as_list(f(args))
  File "/dccstor/aviros_pytorch/pytorch/torch/_inductor/codecache.py", line 479, in __call__
    return self.get_current_callable()(inputs)
  File "/dccstor/aviros_pytorch/pytorch/torch/_inductor/compile_fx.py", line 690, in run
    return compiled_fn(new_inputs)
  File "/dccstor/aviros_pytorch/pytorch/torch/_inductor/cudagraph_trees.py", line 380, in deferred_cudagraphify
    fn, out = cudagraphify(model, inputs, new_static_input_idxs, *args, **kwargs)
  File "/dccstor/aviros_pytorch/pytorch/torch/_inductor/cudagraph_trees.py", line 407, in cudagraphify
    return manager.add_function(
  File "/dccstor/aviros_pytorch/pytorch/torch/_inductor/cudagraph_trees.py", line 1919, in add_function
    return fn, fn(inputs)
  File "/dccstor/aviros_pytorch/pytorch/torch/_inductor/cudagraph_trees.py", line 1739, in run
    out = self._run(new_inputs, function_id)
  File "/dccstor/aviros_pytorch/pytorch/torch/_inductor/cudagraph_trees.py", line 1780, in _run
    return self.run_eager(new_inputs, function_id)
  File "/dccstor/aviros_pytorch/pytorch/torch/_inductor/cudagraph_trees.py", line 1895, in run_eager
    return node.run(new_inputs)
  File "/dccstor/aviros_pytorch/pytorch/torch/_inductor/cudagraph_trees.py", line 605, in run
    out = self.wrapped_function.model(new_inputs)
  File "/dccstor/aviros_pytorch/pytorch/torch/_inductor/compile_fx.py", line 690, in run
    return compiled_fn(new_inputs)
  File "/dccstor/aviros_pytorch/pytorch/torch/_inductor/cudagraph_trees.py", line 370, in deferred_cudagraphify
    return fn(inputs)
  File "/dccstor/aviros_pytorch/pytorch/torch/_inductor/compile_fx.py", line 638, in run
    return model(new_inputs)
  File "/dccstor/aviros_pytorch/pytorch/torch/_inductor/cudagraph_trees.py", line 1738, in run
    assert self.graph is not None, "Running CUDAGraph after shutdown"
AssertionError: Running CUDAGraph after shutdown
```

ani300 committed:
Update on "Add codegen cache"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 aakhundov

[ghstack-poisoned]

eellison commented:
That example works on master. Is it possible you're making it so that `reset_cudagraph_trees` is no longer being called ?

ani300 commented:
`reset_cudagraph_trees` gets called, as I added a print to the call and it shows up... so it must be the initialization later which is not getting called. Let me check what is going on

ani300 commented:
@eellison it only happens when activating the new codegen cache, so I have some more debugging to do

eellison commented on test/inductor/test_cache.py:
```diff
@@ -0,0 +1,151 @@
+if HAS_CUDA and not TEST_WITH_ASAN:
+
+    class TestCodeGenCacheCuda(CacheTestCase):
+        def test_cache_cudagraphs(self):
```
I looked into this. This reason this fails is we create the `CompiledFxGraph`, set its `current_callable` to the [cudagraph'd function](https://github.com/pytorch/pytorch/blob/ea29116794f3beec8ac5a97277f13eb2ab17ad96/torch/_inductor/compile_fx.py#L413) and then store a reference to the compiled graph in memory `cls.cache[key] = compiled_graph`.

Later, when discard the cudagraph with `torch._dynamo.reset()`, that in memory reference is kept alive in `cls.cache`. Next time the key is hit, we return this now-invalid cudagraph function.

This fixes that error:

```
--- a/torch/_inductor/compile_fx.py
+++ b/torch/_inductor/compile_fx.py
@@ -410,8 +410,8 @@ def compile_fx_inner(
             ):
                 boxed_forward_device_index.set(next(iter(compiled_graph.device_idxs)))
 
-            compiled_graph.current_callable = cudagraphify(
-                compiled_graph.get_current_callable(),
+            current_callable = cudagraphify(
+                compiled_graph.compiled_artifact,
```

However, now you an `assertEqual` failure on the output. I don't think this is related to cudagraphs. If you remove `mode=reduce-overhead` from the test and force a File Lookup:
```
--- a/torch/_inductor/codecache.py
+++ b/torch/_inductor/codecache.py
@@ -454,6 +454,9 @@ class FxGraphCache:
                     with open(cg_path, "rb") as f:
                         compiled_graph = pickle.load(f)
 
+                # always force file lookup
+                return compiled_graph
+
                 cls.cache[key] = compiled_graph
```

you get the same error. I don't think we have any test coverage of file look up path.

eellison reviewed (commented):
We don't have any test coverage of the file lookup path right now. 

Also, we are iteratively applying `align_inputs` to the `current_callable()` each time the function is loaded. I think we should be more using `compiled_artifact` more in `compile_fx_inner` than the `current_callable` we are currently using.
https://github.com/pytorch/pytorch/blob/main/torch/_inductor/compile_fx.py#L462-L464

eellison commented:
@ani300 thanks for your work so far. do you have bandwidth to finish this ?

ani300 commented:
@eellison currently not, I'll probably get some bandwidth in around 1-2 weeks, but if you or someone else wants to take it over and bring it to the finish line I'm happy with that too. If I finish everything else I'm working on and can get to it sooner I'll let you know

masnesral commented:
@ani300, checking in with your plans here. @eellison suggested I could take over this PR as part of some onboarding to his team. I'm new to this space, so I won't be efficient. But if you're still a ~week away from getting back, I'm happy to try my hand at getting it over the line... Just wanted to check with you first.

ani300 commented:
@masnesral feel free to take this over, as I'll definitely be busy until mid next week or so. Also feel free to ask any questions

msaroufim commented:
Just trying to understand the remaining work, would it be reasonable to just not make the codecache work with cudagraphs for this PR, merge it and followup in a seperate PR why cudagraphs is failing

CI is mostly green and it'd be a shame not to land this by dev day just because of cudagraphs

eellison commented:
> would it be reasonable to just not make the codecache work with cudagraphs for this PR, merge it and followup in a seperate PR why cudagraphs is failing

I don't think so, because as i said in the review above, we don't have any test coverage of the file lookup path right now.

ani300 commented:
I think getting it working with cudagraphs and adding the tests that @ezyang mentioned a few weeks (months?) ago would be good to get this over the finishing line

masnesral committed:
commandeer PR on "[RFC][inductor] Implement Fx graph caching to improve warm compilation time."


Summary: Implement an on-disk cache to save and reuse compiled FX Graphs.

Test Plan:
* New unit tests exercising saving and load from the cache.
* New unit tests to exercise the cache key calculations.
* Ran several benchmarks to see cache hit and resulting compilation times.

[ghstack-poisoned]

masnesral commented on test/inductor/test_codecache.py:
```diff
@@ -1,9 +1,21 @@
 # Owner(s): ["module: inductor"]
```
Just based on the name, this seemed like the right file to place any new "codecache" unit tests. Lemme know if there's a different convention, however. I was a little confused about the existing stuff in this file because AFAICT there were no unit tests running.

@ani300, you might notice that I didn't preserve any of your previous tests. TBH, I just couldn't figure out what they were doing and what I have here seemed more straightforward (to me). Please help me out if I should actually be carrying over some of the tests you'd written and I'll gladly bring them back.

ani300 replied:
My tests were only testing the behavior of compile model, dynamo reset, compile model again and see if we hit the cache if we expect to. I was missing all the hashing tests that you've added and kwargs ordering etc.

masnesral commented on test/inductor/test_codecache.py:
```diff
@@ -37,3 +49,209 @@ def test_codecache_spawn():
+        Verify that we can populate and load CompiledFxGraphs from the cache.
+        """
+
+        def fn(x, y):
```
Would love input on whether people think we should test compiling/caching more sophisticated code. Also, is it valuable to run an instance of this test with the tensors on cpu? Different dtypes, perhaps?

Chillee replied:
Yeah I would definitely set up tests that have inputs with 1. different strides, 2. different dtypes, 3. different storage offsets, 4. different aliasing of inputs.

eellison replied:
I think we should add an option in the runner that would compile the model, torch._dynamo.reset(), then check that when run again we both hit the cache and accuracy passes. And we should enable that wherever we're checking accuracy right now. 

Any other thoughts on testing @ezyang ?

ani300 replied:
@masneral @eellison my original tests were supposed to be doing just this, compiling, reset, compiling again. This is where I hit all those cudagraph errors.

masnesral replied:
I'll add `torch._dynamo.reset()` to the test(s), but I think @eellison is also proposing to add a new feature to the benchmark runner?

eellison replied:
Right. Would be great to test this on models in ci.

ezyang replied:
NGL, I don't really think any of our existing CI is going to do all that well at catching errors in caching. There are two things you want to test:

1. Cache hits
2. Cache missing when you should

But existing CI isn't going to have a pre-warmed cache to test (1) nor (2), because in CI you mostly want to run everything once, cold.

To test cache hits, I suppose you could add a testing mode where instead of directly running a compiled product, you serialize it out, deserialize it, and check the deserialization still passes tests. But this won't catch problems that occur only when you change process. And running every test twice... well... this doesn't seem like good ROI, except maybe some occasional fuzz test (not even sure if I'd want it master only).

To test cache misses, I suppose you can fuzz this by chucking every compiled product into the cache and checking that hits give the same result. This will be a huge pain to debug if it does fail though (because you need to find exactly the right other test which produced the incompatible product) and I'd imagine most of the time things were different enough that you didn't really hit any nasty edge cases.

I think a more promising approach is to assume that we *are* going to do it wrong, and figure out how to make it as pleasant and easy to debug when we have done it wrong. Stuff like recording provenance where cached products came from, logging when you use a cached product, an easy way to ignore cache, etc.

masnesral commented on torch/_inductor/codecache.py:
```diff
@@ -379,6 +388,163 @@ def write_atomic(path: str, content: Union[str, bytes]) -> None:
+    """
+    See FxGraphCachePickler. Custom reducer to enable pickling TypedStorage.
+    """
+    # The string repr seems to capture the relevant details for hashing, e.g.,
```
For some of the benchmarks I tested, I saw TypedStorage objects in the torch.fx.GraphModule, and those were preventing cache hits when I straightforwardly pickle the whole object without playing this trick (because the _cdata field changes among runs). I don't _really_ know what a TypedStorage is, but from the docs sounds like some lower-level object under Tensor.

I could use some guidance here on the salient data to capture in order to hash safely. I don't _really_ like using the string representation, but to my naive eyes it seems to capture the relevant attributes of a TypedStorage, plus feels kinda future proof in case other fields are added in the future. Perhaps we could ignore the actual value, however (0.0 in the example I show here)??? In my testing, including the value didn't have an impact on cache hits for torchbench benchmarks.

Chillee replied:
Where do you see typedstorage objects in the graphmodule? On the metadata?

masnesral replied:
> I don't know when we would be using a Storage and not a Tensor in compilation 🤔 more information would be helpful here

@eellison, ok the instances of TypedStorage that I'm seeing _are_ in Tensors. I don't understand exactly what's going on, but I've discovered that much. Does it make sense that there would be Tensor objects in the graphmodule? I guess I naively assumed that every Tensor would be a FakeTensor. It seems like I can treat Tensors the same was as FakeTensors for the purposes of computing the hash key, but I think you also have a concern about that approach?

eellison replied:
Constants will appear as attributes of the graph module. Some of the constants are inlined directly into codegen, in which case we'd need to check for complete equality of values. The ones that are not inlined we'd just need to check metadata equality... Maybe this is better handled as a stacked pr / follow up, and bail for now when constants appear ? Up to you.

masnesral commented on torch/_inductor/codecache.py:
```diff
@@ -379,6 +388,163 @@ def write_atomic(path: str, content: Union[str, bytes]) -> None:
+    return (TensorMetadataHolder, (_extract_tensor_metadata(t), t.device))
+
+
+@dataclasses.dataclass
```
I got this idea from here: https://github.com/pytorch/pytorch/blob/main/torch/_inductor/debug.py#L475-L505

masnesral commented on torch/_inductor/codecache.py:
```diff
@@ -379,6 +388,163 @@ def write_atomic(path: str, content: Union[str, bytes]) -> None:
+    Supports caching and reusing compiled Fx graphs.
+    """
+
+    # TODO(slarsen): Investigate whether it's beneficial to store compiled graphs
```
The first impl from @ani300 was also keeping the data in an in-memory cache, but I can't convince myself that it leads to any noticeable perf improvement. Since @eellison notes a problem when keeping the reference alive anyway, I propose we keep the impl simpler and I'll later look into whether having an in-memory component makes any real difference.

masnesral commented on torch/_inductor/graph.py:
```diff
@@ -956,14 +956,13 @@ def compile_to_module(self):
         self.cache_path = path
         self.cache_linemap = linemap
 
-        for name, value in self.constants.items():
```
These are now set by PyCodeCache.load_by_key_path

masnesral commented on torch/_inductor/codecache.py:
```diff
@@ -392,6 +558,7 @@ class CompiledFxGraph:
     device_idxs: Set[int] = field(default_factory=set)
     mutated_inputs: Set[str] = field(default_factory=set)
     mutated_input_idxs: Set[int] = field(default_factory=set)
+    constants: Dict[str, torch.Tensor] = field(default_factory=dict)
```
I discovered we need to keep these constants so that we can add them as module attributes. That's what happens on the "non-cache" code path and I found 2 torchbench benchmarks that were failing with: `CUDA kernel errors might be asynchronously reported at some other API call, so the stacktrace below might be incorrect.`

masnesral commented on torch/_inductor/codecache.py:
```diff
@@ -379,6 +388,163 @@ def write_atomic(path: str, content: Union[str, bytes]) -> None:
+    items: List[Any]
+
+
+class FxGraphHashDetails:
```
@eellison, note that I haven't included the guards in the data that I pickle/hash. I was getting: `'cannot pickle 'weakref.ReferenceType' object'`. I'm sure I can find a fix for that, but I thought it might be worth discussing whether it's necessary? I naively understand guards to capture Tensor shapes and the implementation here captures a lot of info about the tensors. For example:

`TensorMetadataHolder(tensor_metadata=TensorMetadata(shape=torch.Size([128, 64, 1, 1]), dtype=torch.float32, requires_grad=True, stride=(64, 1, 1, 1), memory_format=torch.contiguous_format, is_quantized=False, qparams={}), device=device(type='cuda', index=0))`

Is this enough? Or is my understanding too shallow and there are other kinds of guards?

Chillee replied:
I don't think we need to include guards in the cache key. I think the guards only matter insofar as they affect what the graph looks like. 

I think for symbolic shapes it's a little more subtle, but I think the key we're caching on shouldn't be the guards but the symbolic shape env itself.

eellison replied:
I think for symbolic shapes the key should be the symbolic shape guards. They represent when the inductor graph is valid for reuse.

masnesral replied:
Ok, I'm not grokking what I need to capture to make sure symbolic shapes are handled correctly. Forgive me if I'm way off, but I do notice that the Tensor metadata we capture in the existing implementation can look like this (if it's at all relevant):

`TensorMetadata(shape=torch.Size([s0, s1]), dtype=torch.bfloat16, requires_grad=False, stride=(s1, 1), memory_format=torch.contiguous_format, is_quantized=False, qparams={})`

eellison replied:
A proposal for a symbolic shape scheme:
- we should hash to make sure the same symbolic shape relationships appear in the same way e.g. `[s0, s1]` is different than `[s0, s0]`
- Serialize the guards we've added for each symbolic shape. 
- After you have a cache hit, at runtime test that the guards we've serialized pass for the runtime symbolic shapes

masnesral replied:
See my TODO below. I'm wondering if it's maybe possibly sufficient to just consider shape_env.var_to_range

masnesral replied:
synced with @eellison; the plan is to change this PR to handle static only and then do this proposal in a next step

masnesral commented:
@Chillee FYI

masnesral commented:
Question (maybe for @eellison and/or @ezyang): I've been testing with a bunch of torchbench benchmarks. My (probably naive) expectation is that if I run the benchmark once, then I should see complete cache hits on a 2nd run. But there are some occasional and inconsistent misses. For example, I might run `basic_gnn_gcn` 3 times in a row with cache hits and then see a miss on the 4th run (or not). I've tracked it down to the `memory_format` field of a random tensor. Sometimes it's `memory_format=torch.contiguous_format` and sometimes `memory_format=torch.contiguous_format`. Any insight? Is this just some builtin variability of some of the benchmarks? I assume it's not safe to cache if the memory format is different.

Chillee commented on test/inductor/test_codecache.py:
```diff
@@ -37,3 +49,209 @@ def test_codecache_spawn():
+
+    def test_hash_typed_storage(self):
+        """
+        Test hashing (pickling) TypedStorage objects.
```
I'm a bit confused about where `TypedStorage` shows up? Perhaps I'll figure it out reading through more of the code.

masnesral replied:
See larger comment below on this topic.

Chillee commented:
Do you have any preliminary performance numbers?

eellison reviewed (commented):
Looks good ! A few things we should think through and warrant discussion... 

1) Cache validity. The goal initially is for a model that was previously run to hit the cache successfully. We are not trying for maximal reuse. 

For static shape tensors, we should be able to use whatever cache we end up at https://github.com/pytorch/pytorch/pull/110119/files#r1343291742.  

I don't **think** we need to guard on aliasing relationships since this should be post-functionalization.

For dynamic shapes, I think it would make sense to extract the existing dynamic shape guards and reuse them. The guards represent when it is safe to reuse a compilation which is exactly what we need here. 

It is possible for us to add more guards in inductor compilation, so I don't think we can just compare the set of guards that exists with what exists in the cache key. Potentially something along the lines of caching the dynamic shape guards before inductor compilation begins, then serializing the additional ones that were added with the codecache and doing a runtime check.

2) Testing. We should have tests for the basic stuff - different shapes, dtype, devices, striding, aliasing, symbolic shapes. 

But it's going to be hard to maintain this if it's only run in a small test suite. We should also extend this to the `accuracy` runner for benchmarks/common. 

Not for this pr, but it would be cool to figure out a way to re-use the cache for different cuda devices.

Chillee commented on torch/_inductor/codecache.py:
```diff
@@ -379,6 +388,163 @@ def write_atomic(path: str, content: Union[str, bytes]) -> None:
+                else:
+                    self.fx_kwargs[k] = fx_kwargs[k]
+
+        self.torch_version = torch.__version__
```
Oh that's pretty smart. Yeah I think that makes sense.

masnesral replied:
> Should we also cache on the source code in inductor ?

@eellison, I assume you mean "hash" the source code in inductor and not "cache"? If so, where do I get that. And is it different from the code that's already captured from the GraphModule argument?

eellison replied:
Yes, hash. Triton has similar code here: https://github.com/openai/triton/blob/020f43d5a37c2f1cce5c70fcda17ec57b000755c/python/triton/runtime/jit.py#L127

We'd be hashing on inductor source code so if you make changes you dont get a stale compilation.

masnesral replied:
I totally misunderstood what you meant by "inductor code". Thanks for the pointer.

masnesral commented:
> Do you have any preliminary performance numbers?

@Chillee, I assume you mean, e.g., just run some benchmarks on my devgpu? If so, do you have any specific benchmarks / flags in mind that would be illuminating? And what's the most relevant number? End-to-end run time of a benchmark or are you just asking about the cumulative compile time? For the `resnet18` kernel you gave me, it takes ~4s for `fx_codegen_and_compile`. The current impl does not serialize the "compiled_artifact" field (see comments in code), and it looks like it takes ~0.5s to recreate it after a cache hit, for a total savings of ~3.5s. So total time for `torch.compile()` + execution for that example code is something like 17.5s with a cold cache and 14s with a cache hit. I've seen some torchbench benchmarks with longer compile times, so they might see more impressive numbers....

Chillee commented:
@masnesral How did you get those timings? When I do this, resnet18 seems to take 33 seconds to compile for me, 25 seconds of which is spend in `fx_codegen_and_compile`.

```
import random
import torch
import torch.utils.benchmark as benchmark
import torch.nn.functional as F
import functools
from itertools import product
import time

torch.set_default_device('cuda')
from torchvision.models import resnet18

model = torch.compile(resnet18())
begin = time.time()
model(torch.randn(1, 3, 224, 224))
print(time.time() - begin)
```

masnesral commented:
@Chillee, we were using 

```
with torch.no_grad():
    model(torch.randn(1, 3, 224, 224))
```

But I still get the same basic results using your sample here:

```
(pytorch-3.10) slarsen@devgpu024:pytorch (codegen_cache) $ TORCHINDUCTOR_FX_GRAPH_CACHE=0 TORCH_LOGS=+torch._inductor.codecache,+torch._inductor.compile_fx python /home/slarsen/horace2.py
/home/slarsen/local/pytorch/torch/_inductor/compile_fx.py:129: UserWarning: TensorFloat32 tensor cores for float32 matrix multiplication available but not enabled. Consider setting `torch.set_float32_matmul_precision('high')` for better performance.
  warnings.warn(
[2023-10-03 08:04:06,494] [0/0] torch._inductor.compile_fx: [INFO] Step 3: torchinductor compiling FORWARDS graph 0
[2023-10-03 08:04:10,938] [0/0] torch._inductor.compile_fx: [DEBUG] FX codegen and compilation took 4.444s
[2023-10-03 08:04:10,939] [0/0] torch._inductor.compile_fx: [INFO] Step 3: torchinductor done compiling FORWARDS graph 0
14.625285387039185
```

```
$ TORCHINDUCTOR_FX_GRAPH_CACHE=1 TORCH_LOGS=+torch._inductor.codecache,+torch._inductor.compile_fx python /home/slarsen/horace2.py
[2023-10-03 08:02:50,054] [0/0] torch._inductor.codecache: [DEBUG] fx graph cache: on-disk miss for key fihwjtavgh7lffl6lvcq5om25us65tp3diqz4owdc6kiysn73v25
/home/slarsen/local/pytorch/torch/_inductor/compile_fx.py:129: UserWarning: TensorFloat32 tensor cores for float32 matrix multiplication available but not enabled. Consider setting `torch.set_float32_matmul_precision('high')` for better performance.
  warnings.warn(
[2023-10-03 08:02:50,055] [0/0] torch._inductor.compile_fx: [INFO] Step 3: torchinductor compiling FORWARDS graph 0
[2023-10-03 08:02:54,946] [0/0] torch._inductor.codecache: [DEBUG] fx graph cache: wrote on-disk entry for key fihwjtavgh7lffl6lvcq5om25us65tp3diqz4owdc6kiysn73v25
[2023-10-03 08:02:54,946] [0/0] torch._inductor.compile_fx: [DEBUG] FX codegen and compilation took 4.922s
[2023-10-03 08:02:54,947] [0/0] torch._inductor.compile_fx: [INFO] Step 3: torchinductor done compiling FORWARDS graph 0
15.124929428100586
```

```
(pytorch-3.10) slarsen@devgpu024:pytorch (codegen_cache) $
(pytorch-3.10) slarsen@devgpu024:pytorch (codegen_cache) $ TORCHINDUCTOR_FX_GRAPH_CACHE=1 TORCH_LOGS=+torch._inductor.codecache,+torch._inductor.compile_fx python /home/slarsen/horace2.py
[2023-10-03 08:03:31,501] [0/0] torch._inductor.codecache: [DEBUG] fx graph cache: on-disk hit for key ffbquqm7wn54ghnsfqtobshxza3kep5mt667xml56eiy2t3ecqrz
[2023-10-03 08:03:31,501] [0/0] torch._inductor.compile_fx: [DEBUG] FX codegen and compilation took 0.029s
[2023-10-03 08:03:31,502] [0/0] torch._inductor.compile_fx: [INFO] Step 3: torchinductor done compiling FORWARDS graph 0
11.05607533454895
```

ani300 commented:
> Question (maybe for @eellison and/or @ezyang): I've been testing with a bunch of torchbench benchmarks. My (probably naive) expectation is that if I run the benchmark once, then I should see complete cache hits on a 2nd run. But there are some occasional and inconsistent misses. For example, I might run `basic_gnn_gcn` 3 times in a row with cache hits and then see a miss on the 4th run (or not). I've tracked it down to the `memory_format` field of a random tensor. Sometimes it's `memory_format=torch.contiguous_format` and sometimes `memory_format=torch.contiguous_format`. Any insight? Is this just some builtin variability of some of the benchmarks? I assume it's not safe to cache if the memory format is different.

I assume there is a typo in this comment? You mention contiguous format twice

ani300 commented on torch/_inductor/codecache.py:
```diff
@@ -379,6 +388,163 @@ def write_atomic(path: str, content: Union[str, bytes]) -> None:
+    serialized_data = FxGraphCachePickler.dumps(details)
+    return (
+        "f"
+        + base64.b32encode(hashlib.sha256(serialized_data).digest())[:51]
```
this is to keep the name in the filesystem short, and I copied it from the Inductor code cache, so it is consistent with that one

masnesral commented:
> I assume there is a typo in this comment? You mention contiguous format twice

@ani300, yes thank you. I meant `memory_format=torch.channels_last` and `memory_format=torch.contiguous_format`

masnesral commented on torch/_inductor/codecache.py:
```diff
@@ -379,6 +388,163 @@ def write_atomic(path: str, content: Union[str, bytes]) -> None:
+
+@dataclasses.dataclass
+class TensorMetadataHolder:
+    tensor_metadata: TensorMetadata
```
@eellison Can you elaborate? Do you mean that we shouldn't reuse `TensorMetadata` and `_extract_tensor_metadata` and instead just build it fresh here?

eellison replied:
Right, for instance, it doesn't handle `device` and it's default handling of symbolic shapes (as you commented elsewhere) won't be sufficient

masnesral replied:
Ok, took a crack at re-rolling this. Resolving this convo.

eellison commented:
> example, I might run basic_gnn_gcn 3 times in a row with cache hits and then see a miss on the 4th run (or not). I've tracked it down to the memory_format field of a random tensor. Sometimes it's memory_format=torch.contiguous_format and sometimes memory_format=torch.contiguous_format

Sounds like non-determinism elsewhere in the compiler, which is its own bug.

masnesral commented on torch/_inductor/codecache.py:
```diff
@@ -379,6 +388,163 @@ def write_atomic(path: str, content: Union[str, bytes]) -> None:
+
+        try:
+            key = compiled_fx_graph_hash(fx_args, fx_kwargs)
+        except Exception as ex:
```
@eellison Was just trying to be defensive in case the hashing fails. I have not seen it fail in my recent testing (running as many benchmarks as I can), however. I can certainly remove if you think it's better to crash.

eellison replied:
Let's remove this then, especially since this isn't on by default yet anyway.

masnesral commented on torch/fx/graph_module.py:
```diff
@@ -106,7 +106,9 @@ def _format_import_block(globals: Dict[str, Any], importer: Importer):
-    return "\n".join(import_strs)
+    # Sort the imports so we have a stable import block that allows us to
+    # hash the graph module and get a consistent key for use in a cache.
+    return "\n".join(sorted(import_strs))
```
@eellison I thought it was just the behavior of Python's set. I don't think set insertion preserves anything.

eellison replied:
where as we using set ? maybe we should have a linter rule in torch/_inductor to never use set(), and always use an ordered set() variant

masnesral replied:
Just a few lines above on line 106: `import_strs: Set[str] = set()`
Sure; I can change the usage here and I'm pretty certain it will work fine. But what ordered set variant? Do you know of one? I was surprised to find that collections lib, for example, doesn't have one. Googling and ChatGPT'ing point to rando third-party libs that I don't think we're currently using anywhere this codebase. I could do it with a dict?

ezyang commented:
in case you miss it: https://github.com/pytorch/pytorch/pull/103453#discussion_r1346393445

masnesral committed:
reviewer comments on "[inductor] Implement Fx graph caching to improve warm compilation time."


Summary: Implement an on-disk cache to save and reuse compiled FX Graphs.

Test Plan:
* New unit tests exercising saving and load from the cache.
* New unit tests to exercise the cache key calculations.
* Ran several benchmarks to see cache hit and resulting compilation times.

[ghstack-poisoned]

masnesral commented on torch/_inductor/codecache.py:
```diff
@@ -379,6 +386,240 @@ def write_atomic(path: str, content: Union[str, bytes]) -> None:
+@dataclasses.dataclass
+class TensorMetadata:
+    """
+    The Tensor metadata relevant when hashing FxGraph cache keys.
```
Please let me know if there are other tensor fields that are relevant to hashing

masnesral commented on torch/_inductor/codecache.py:
```diff
@@ -379,6 +386,240 @@ def write_atomic(path: str, content: Union[str, bytes]) -> None:
+    """
+    Extract the TensorMetadata of a tensor.
+    """
+    # Note the order here matters since is_contiguous() can be True
```
I figured out why I was occasionally seeing different memory formats on consecutive runs. It's possible for is_contiguous to match > 1 of the memory formats options. So the order here matters and this iterable should not be a set.

masnesral commented on torch/_inductor/codecache.py:
```diff
@@ -379,6 +386,240 @@ def write_atomic(path: str, content: Union[str, bytes]) -> None:
+    """
+    # See GraphLowering. We expect any Tensors to be small inlined constants.
+    # For these, we hash on the metadata _and_ the actual values.
+    assert len(t.shape) == 0 or (len(t.shape) == 1 and t.shape[0] <= 128)
```
This part isn't quite right. @eellison tells me there are some tensors for which it's safe to consider only their metadata and others (inlined constants) where we should also consider the values. This 128 in the assert is totally arbitrary.

masnesral committed:
tensor constants on "[inductor] Implement Fx graph caching to improve warm compilation time."


Summary: Implement an on-disk cache to save and reuse compiled FX Graphs. This implementation does not handle tensors with symbolic shapes. This needs to be done in a follow-up PR.

Test Plan:
* New unit tests exercising saving and load from the cache.
* New unit tests to exercise the cache key calculations.
* Ran several benchmarks to see cache hit and resulting compilation times.

[ghstack-poisoned]

eellison commented on torch/_inductor/codecache.py:
```diff
@@ -458,13 +469,17 @@ def _reduce_tensor(t):
+    # Large constannts are effectively treated as inputs and we consider only
+    # their metadata.
+    metadata = extract_tensor_metadata(t)
+    if len(t.shape) == 0 or (len(t.shape) == 1 and t.shape[0] <= 8):
```
Can you factor out the logic in torch/_inductor/graph.py about whether or not we are inlining the constants and call that here ? It's fragile to duplicate the logic.

masnesral replied:
@eellison, ah good catch. Thanks; meant to ask about this, but brain farted. I wondered if the logic might already be duplicated because this doesn't look like the place where the inlining actually happens. I found this: https://github.com/pytorch/pytorch/blob/main/torch/_inductor/lowering.py#L2217 and it actually looks slightly different (the check for float or int). Which is actually the better source of truth here?

eellison replied:
That's a good find! But I don't if that lowering matters since the inputs will be represented in the string of the graph. I think it's just the attr constants that would matter. 

TBh i'm not sure when we would generate `torch.tensor` or `scalar_tensor` - I would have thought those would get replaced by `lift_fresh`. Maybe @ezyang knows since he touched it somewhat recently ? https://github.com/pytorch/pytorch/commit/39ba2e6226eb78ec2e73b0636f6cb599eeff3e62

In any case I'm pretty sure you can ignore that lowering for constant hashing purposes since the input values will be represented in the graph.

ezyang replied:
`scalar_tensor` gets generated if the inputs to torch.tensor have SymInt in them. You can't use lift fresh because the tensor contents are actually dynamic.

eellison commented on torch/_inductor/codecache.py:
```diff
@@ -379,6 +386,254 @@ def write_atomic(path: str, content: Union[str, bytes]) -> None:
+    ]
+
+    memory_format = None
+    for fmt in memory_formats:
```
you can replace this loop with `torch._prims_common.suggest_memory_format`, although it will default to contiguous, so maybe after invoking that api check `suggested_memory_format = suggested_memory_format if t.is_contiguous(suggested_memory_format) else None`

eellison commented on torch/_inductor/codecache.py:
```diff
@@ -379,6 +386,254 @@ def write_atomic(path: str, content: Union[str, bytes]) -> None:
+    return TensorMetadata(
+        dtype=t.dtype,
+        shape=t.shape,
+        stride=t.stride(),
```
some tensors dont have strides - you can test this with a sparse tensor - maybe safer to do `t.stride() if t.layout == torch.strided_layout else None)`

eellison commented on torch/_inductor/codecache.py:
```diff
@@ -346,6 +351,8 @@ def get_hash(content: Union[str, bytes], extra: str = "", hash_type: str = "code
         return code_hash(content, extra)
     if hash_type == "cubin":
         return code_hash(repr(content))
+    if hash_type == "cg":
```
mind explaining this ?

masnesral replied:
@eellison it's taken from @ani300's original diff. There are some existing utilities in this file to read/write the entries on disk (see `get_path`, `get_hash`, `write`). I think it basically boils down to an attempt to make the `write` method sorta generic for different cache types in that it computes the hash differently based on the type, supplied as this hash_type arg. It doesn't work all that cleanly here. I can try to improve, but honestly I think the FxGraphCache is going to need its own separate read/write logic anyway in a follow-up PR because your proposal requires that we handle multiple entries per key.

eellison commented on torch/_inductor/codecache.py:
```diff
@@ -379,6 +386,254 @@ def write_atomic(path: str, content: Union[str, bytes]) -> None:
+    requires_grad: bool
+    storage_offset: int
+    memory_format: Optional[torch.memory_format]
+    is_quantized: bool
```
For completeness, let's also add: `is_conj()`, `is_neg()` and `dense_dim()`, `sparse_dim()`, `is_coalesced()` in the case of sparse tensor:

https://github.com/pytorch/pytorch/blob/main/torch/_subclasses/meta_utils.py#L55-L63

eellison commented on test/inductor/test_codecache.py:
```diff
@@ -37,3 +53,245 @@ def test_codecache_spawn():
+        """
+        Test the handling of small vs. large tensor constants.
+        """
+        data = FxGraphCachePickler.dumps(torch.tensor([i for i in range(9)]))
```
Maybe could you add some that explicitly invoke inductor ? 

e.g. with something like

```
        def fn():
            return (torch.tensor([i for i in range(8)], device="cuda") + 10,)

        fn_fx = make_fx(fn)()
        fn_compiled = compile_fx_inner(fn_fx, [])
```

and changing the generated constant and see if we do a new recompilation

masnesral replied:
Ok, for my education why do you suggest to explicitly invoke inductor for this particular test vs. a torch.compile?

eellison reviewed (commented):
looks great ! I think worth landing especially since this isn't on by default yet. A couple bit todos are :

- dynamic shapes
- better logging / visibility as ed suggested
- some sort of more comprehensive testing
- turn on by default


Before we land, can you bring in the tests that @ani300 had on his original pr ? those are additional coverage for free.

masnesral commented:
> Before we land, can you bring in the tests that @ani300 had on his original pr ? those are additional coverage for free

@eellison, I think those won't work "out of the box" now because I'm not doing the in-mem caching. Also, I removed the randomness as requested on earlier revs. But I think I'm basically testing exactly the same thing. BUT: I am missing the testing of both functions and torch.nn.Module, so I shall add that part in.

eellison commented:
in-mem storage should just be an optimization on top of disk level storage. why wouldn't it work ?

masnesral commented:
> in-mem storage should just be an optimization on top of disk level storage. why wouldn't it work ?

@eellison, I probably misspoke. It looks like there was an attempt to explicitly test functionality of in-mem, but AFAICT it wasn't fully complete. See `activate_unique_cache()` for example, some comments about cache collisions, and reviewer requests to remove the rng. So I'd just need to clean them up a little. But I thought the current tests I have are a superset of those previous tests. Except perhaps checking `torch._dynamo.utils.compilation_time_metrics`? I thought the use of mocks to explicitly check the number of cache hits and misses is more direct.

eellison commented:
The current tests do not test `reduce-overhead`, for example https://github.com/pytorch/pytorch/pull/103453/commits/2b1cb6ec3c769e85b0678f09cadd1666453d4a1f#diff-68d0993a1a3fae1da3b1ffb633dcb5566f3f6afca87bef3488482cd36ba57c09R63-R144. 

I don't think the tests are a strict superset and having additional tests can't hurt. The core of the tests are checking whether or not a cache hit vs cache miss occurred. We should do be able to do something similar. See, torch._dynamo.counters and use/reset of them here: https://github.com/pytorch/pytorch/blob/main/test/inductor/test_pattern_matcher.py#L59-L65.

masnesral committed:
comments and counters on "[inductor] Implement Fx graph caching to improve warm compilation time."


Summary: Implement an on-disk cache to save and reuse compiled FX Graphs. This implementation does not handle tensors with symbolic shapes. This needs to be done in a follow-up PR.

Test Plan:
* New unit tests exercising saving and load from the cache.
* New unit tests to exercise the cache key calculations.
* Ran several benchmarks to see cache hit and resulting compilation times.

[ghstack-poisoned]

masnesral committed:
lint on "[inductor] Implement Fx graph caching to improve warm compilation time."


Summary: Implement an on-disk cache to save and reuse compiled FX Graphs. This implementation does not handle tensors with symbolic shapes. This needs to be done in a follow-up PR.

Test Plan:
* New unit tests exercising saving and load from the cache.
* New unit tests to exercise the cache key calculations.
* Ran several benchmarks to see cache hit and resulting compilation times.

[ghstack-poisoned]

masnesral commented:
@pytorchbot merge

kit1980 commented:
@pytorchbot revert -m "The newly added test fails on internal builds" -c ghfirst

masnesral commented:
> ani300 your PR has been successfully reverted.

Looking into it

masnesral committed:
hopefully fix tests on "[inductor] Implement Fx graph caching to improve warm compilation time."


Summary: Implement an on-disk cache to save and reuse compiled FX Graphs. This implementation does not handle tensors with symbolic shapes. This needs to be done in a follow-up PR.

Test Plan:
* New unit tests exercising saving and load from the cache.
* New unit tests to exercise the cache key calculations.
* Ran several benchmarks to see cache hit and resulting compilation times.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

masnesral committed:
Update on "[inductor] Implement Fx graph caching to improve warm compilation time."


Summary: Implement an on-disk cache to save and reuse compiled FX Graphs. This implementation does not handle tensors with symbolic shapes. This needs to be done in a follow-up PR.

Test Plan:
* New unit tests exercising saving and load from the cache.
* New unit tests to exercise the cache key calculations.
* Ran several benchmarks to see cache hit and resulting compilation times.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

Chillee reviewed (approved):
Stamping for reland

masnesral commented:
@pytorchbot merge

masnesral committed:
Update on "[inductor] Implement Fx graph caching to improve warm compilation time."


Summary: Implement an on-disk cache to save and reuse compiled FX Graphs. This implementation does not handle tensors with symbolic shapes. This needs to be done in a follow-up PR.

Test Plan:
* New unit tests exercising saving and load from the cache.
* New unit tests to exercise the cache key calculations.
* Ran several benchmarks to see cache hit and resulting compilation times.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

masnesral committed:
lint on "[inductor] Implement Fx graph caching to improve warm compilation time."


Summary: Implement an on-disk cache to save and reuse compiled FX Graphs. This implementation does not handle tensors with symbolic shapes. This needs to be done in a follow-up PR.

Test Plan:
* New unit tests exercising saving and load from the cache.
* New unit tests to exercise the cache key calculations.
* Ran several benchmarks to see cache hit and resulting compilation times.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

masnesral commented:
@masnesral has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D50255289).

masnesral commented:
@pytorchbot merge
